src/QsCompiler/CompilationManager/DataStructures.cs

```diff
@@ -41,69 +50,374 @@ internal class CodeLine
+            {
+                if (delim == -1 || delim == text.Length)
+                {
+                    delimiterBuilder.Add(delim);
+                    continue;
+                }
+
+                var inputDelimiter = text[delim].ToString();
+                // If the input is a " preceded by a $ than the input needs to be updated to $"
+                if (delim > 0 && inputDelimiter == "\"" && text[delim - 1] == '$')
+                {
+                    inputDelimiter = "$\"";
+                }
+
+                bool validDelimiter;
+                (this.EndingStringContext, validDelimiter) = MoveToNextState(this.EndingStringContext, inputDelimiter);
+                if (!validDelimiter)
+                {
+                    // If the erroneous delimiter is $", mark the $ as the error and treat the " as a normal string delimiter
+                    if (inputDelimiter == "$\"")
+                    {
+                        errorDelimiterBuilder.Add(delim - 1); // Backup to the $ character
+                        delimiterBuilder.Add(delim);
+                    }
+                    else
+                    {
+                        errorDelimiterBuilder.Add(delim);
+                    }
+                }
+                else
+                {
+                    delimiterBuilder.Add(delim);
+                }
+            }
+
+            var lineLength = text.Length - this.LineEnding.Length;
+            // if there is a comment
+            if (commentStart >= 0)
+            {
+                this.WithoutEnding = text.Substring(0, commentStart);
+                var commentStr = text.Substring(commentStart, lineLength - commentStart).Trim();
+                var isDocComment = string.IsNullOrWhiteSpace(this.WithoutEnding) && (commentStr.Length - commentStr.TrimStart('/').Length == 3);
+                this.EndOfLineComment = isDocComment ? null : commentStr.Substring(2);
+            }
+            // else there is no comment
+            else
+            {
+                this.WithoutEnding = text.Substring(0, lineLength);
+                this.EndOfLineComment = null;
+            }
+
+            this.StringDelimiters = delimiterBuilder.ToImmutable();
+            this.ErrorDelimiterPositions = errorDelimiterBuilder.ToImmutable();
+            this.Indentation = indentation;
+            this.ExcessBracketPositions = excessBrackets.ToImmutableArray();
+            ScopeTracking.VerifyStringDelimiters(text, this.StringDelimiters);
+            ScopeTracking.VerifyExcessBracketPositions(this, excessBrackets);
         }
 
-        internal CodeLine SetStringDelimiters(IEnumerable<int> newStringDelimiters)
+        /// <summary>
+        /// Computes the location of the string delimiters within a given text.
+        /// </summary>
+        private static IEnumerable<int> ComputeStringDelimiters(string text, ref StringContext stringContext, out int commentIndex, out IEnumerable<int> errorDelimiters)
```

Comment:
This function is a seems to play a role of tokenizer in analyzing the input. As written, it does a lot of scanning over the line (i.e. repeatable 'indexof' with different arguments. A typical DFA approach would consume less CPU time.

Reply:
I've reworked the function so that it should only perform a single pass through the text now.